### PR TITLE
Declare the product version once, and pin every reader to it (#3222)

### PR DESCRIPTION
**Pull request by erikdarlingdata**
Fixes #3222. (The repo's default branch is `main`, so a closing keyword does not fire on a merge to `dev` — #3222 needs closing explicitly.)

## What changed

`Directory.Build.props` at the repository root now holds the one `<Version>` (and the `InformationalVersion` suffix suppression that is the other half of the same fact). All 22 projects inherit it — measured with `dotnet msbuild -getProperty:Version`, not assumed.

The six per-project `<Version>` elements are gone. So are the nine hand-set `AssemblyVersion` / `FileVersion` / `InformationalVersion` elements in the three `deprecated/` projects, which #2113 deleted from the shipped three and never reached.

All nine version reads now name the declaring file: `check-version-bump.yml` (PR side and main side), `nightly.yml` ×2, `build.yml` ×2, `build-lite.cmd`, `build-dashboard.cmd`, `package-release.cmd`.

`Directory.Build.props` is added to `build.yml`'s `root:` path filter. Without it a release bump would touch one repo-root file matching no area filter, and nothing would build.

Each of those nine reads now **fails its own step when it reads nothing**. `ProductVersionDeclarationTests` holds the invariant: one declaration, no hand-set derived property, every reader naming the declaring file, every reader guarded, and every text reader's own pattern agreeing with the XML read.

`UpdateCheckService`'s comment and the release-checklist skill both asserted things this change makes false, and are corrected.

## The issue undercounted the consumers: two named, nine present

The issue named `check-version-bump.yml` and `nightly.yml:114`. Grepping for the literal `<Version>` finds only the text-scraping readers and misses every XML one, so sweeping by *path* instead turns up nine reads across five files — including `build.yml:394`/`:984`, which the issue does not mention, and three root `.cmd` scripts.

Two of those `.cmd` reads pointed at `Dashboard\Dashboard.csproj`, **a path that has not existed since #1612**. `build-dashboard.cmd` has no guard on the result, so it was already producing an empty version silently. That is this defect class already fired and sitting in the tree, and it is why the pin scans the scripts and not just the workflows.

## The gate was certifying a version its own file did not stamp

Read out of built binaries, on the base commit:

| assembly | AssemblyVersion | InformationalVersion |
|---|---|---|
| `PerformanceMonitorDashboard.dll` | **3.3.0.0** | **3.3.0**+sha |
| `Installer.Core.dll` | 3.3.0.0 | 3.3.0+sha |
| `PerformanceMonitorInstaller.dll` | 3.3.0.0 | 3.3.0+sha |

`deprecated/Dashboard/Dashboard.csproj` declared `<Version>3.6.0</Version>` while its binary stamped `3.3.0.0`, because the hand-set derived properties beside it won the stamp. That file is what `check-version-bump.yml` took its judgement from — so the release gate's "version bumped to 3.6.0" was certified from a project whose artifact was three releases behind. #2113's exact defect, live, in the gate's own source of truth.

## Stamped versions, before and after

Read out of the binaries with a metadata reader, not inferred from the props file:

| assembly | before | after |
|---|---|---|
| `PerformanceMonitor.Darling.Service.dll` | 3.6.0.0 / 3.6.0 | **unchanged** |
| `PerformanceMonitor.Darling.Viewer.dll` | 3.6.0.0 / 3.6.0 | **unchanged** |
| `PerformanceMonitorLite.dll` | 3.6.0.0 / 3.6.0 | **unchanged** |
| `PerformanceMonitorDashboard.dll` | 3.3.0.0 / 3.3.0+sha | 3.6.0.0 / 3.6.0 |
| `Installer.Core.dll` | 3.3.0.0 / 3.3.0+sha | 3.6.0.0 / 3.6.0 |
| `PerformanceMonitorInstaller.dll` | 3.3.0.0 / 3.3.0+sha | 3.6.0.0 / 3.6.0 |
| `PerformanceMonitor.Common.dll` | 1.0.0.0 / 1.0.0+sha | 3.6.0.0 / 3.6.0 |
| `Darling.Tests.dll` | 1.0.0.0 / 1.0.0+sha | 3.6.0.0 / 3.6.0 |

**The three shipped apps stamp exactly what they stamped before.** Everything else moves onto the product version, and every `InformationalVersion` loses its source-revision suffix because the suppression is now tree-wide.

## Blast radius of the root props file

MSBuild auto-imports it into every project below it, so the reach is all 22 — the four test projects and the whole `deprecated/` subtree included. That is the trade taken deliberately: an explicit `<Import>` is a convention a new project can forget, which is the shape of the defect being replaced.

Nothing reads a library's or a test assembly's own version to decide anything. `UpdateCheckService` reads the **entry** assembly (its comment claiming `PerformanceMonitor.Common.dll` "carries no `<Version>` and defaults to 1.0.0.0" is now false and is rewritten). `DarlingCliCommands.ProductVersion()` reads the service assembly, which declared a version before and inherits the same one now. The Velopack packs and every artifact name come from the workflow variable, which is unchanged in value — `vpk pack -v $env:VERSION`, where `VERSION` is the workflow output. The channel names (`lite`, `darlingviewer`) are literals and no project property reaches them, so the issue body's question about Velopack channel naming is answered: it is not affected.

## The bug I introduced writing this, and the pin that now holds it

The first draft put the rationale comment above the `PropertyGroup`. The four XML readers select an element and were fine; the three **text** readers take the first match in the file, so all three came back with the word `from` out of the comment prose. Every one of those reads only names an artifact, so nothing goes red — the release tarball is just named `PerformanceMonitorDarling-linux-x64- from.tar.gz`.

The `PropertyGroup` now precedes all prose, and `EveryShippedTextReadPattern_AgreesWithTheXmlRead` extracts each shipped read's **own** pattern from the workflow and script source and requires it to agree with the XML read. Reverting the ordering fails it.

## An empty read was invisible on a pull request

Worth stating because it makes the guards non-obvious. `build.yml`'s version read **executes** on a pull request — its only guard is the docs fast path, which this diff cannot engage. But every consumer of the value is `if: github.event_name == 'release'`, and the step never printed it. So a green PR proved the read did not *throw*; whether it returned anything first mattered on a release, which is the worst place to find out.

The release gate is worse than silent rather than merely silent: two empty reads compare **equal**, so an unguarded pair fails the gate for the wrong reason, while one empty side passes it for the wrong reason.

So each read now fails its step on an empty value — `throw` in the pwsh steps, a `||` exit in the bash steps, `exit /b 1` in the scripts. This covers what no test of the readers' *paths* can: a read aimed at the right file still yields empty if the property is renamed, the file is malformed, or a comment gets in front of the element. The bash guard was verified by executing the shipped lines against the real file (resolves `3.6.0`) and against a props file with no version (exits 1).

Guarding without an instrument would repeat the exact error this issue is about, so `EveryVersionRead_FailsItsStepOnAnEmptyResult` holds it.

## Pins, and the mutation table

Baseline: **24 tests, Failed: 0, Errors: 0, Not Run: 0.** Run by compiling the actual `ProductVersionDeclarationTests.cs` and `RepoFile.cs` into a `net10.0` console harness against real xunit.v3 3.2.2, because `Darling.Tests` is `net10.0-windows` and cannot execute on macOS. CI is the arbiter for the real run.

The violation routes were enumerated rather than sampled — reading assertions finds a wrong one, only enumerating routes finds a missing one. That is how the shadowing route (M5) got an assertion: a second `Directory.Build.props` in a subdirectory declaring *nothing* takes the version from its whole subtree while adding no declaration for a count to catch.

| # | mutation | expected red | result |
|---|---|---|---|
| M1 | a second `<Version>` reappears in Lite's csproj | `ExactlyOneMsBuildProperty…` | RED (1 failed) |
| M2 | a hand-set `<AssemblyVersion>` reappears in the gate's file | `NoProjectHandSetsADerived…` | RED (1 failed) |
| M3 | `nightly.yml`'s read points back at a project file | `EveryVersionRead_Names…` | RED (1 failed) |
| M4 | prose naming the property moves above the element | `EveryShippedTextReadPattern…` | RED (1 failed) |
| M5 | a shadowing `Directory.Build.props`, declaring nothing | `ExactlyOneMsBuildProperty…` | RED (1 failed) |
| M6 | the declared value stops being a version | `ExactlyOneMsBuildProperty…` | RED (1 failed) |
| I1 | the workflow **glob** is crippled to match nothing | the non-vacuity floors | RED (2 failed) |
| I2 | the detector's line **predicate** recognises no read | the non-vacuity floors | RED (11 failed) |
| I3 | the path **matcher** accepts everything | the detector theory | RED (5 failed) — and the repo scan **passed** |
| I4 | the detector loses its text-scrape half | `reads >= 9` floor | RED (1 failed) |
| I5 | crippled glob + one test's floors deleted | — | RED — a *second* test floors the same glob |
| I5b | crippled glob + **all three** floors deleted | — | **GREEN, 17/17** |
| I6 | element walk becomes a substring count | — | GREEN (see below) |
| G1 | a guard is deleted from `nightly.yml`'s artifact-naming read | `EveryVersionRead_FailsItsStep…` | RED (1 failed) |
| G2 | a guard is deleted from `build.yml`'s release-tarball read | `EveryVersionRead_FailsItsStep…` | RED (1 failed) |
| G3 | the guard detector credits **everything** | the guard theory | RED (2 failed) — and the guard scan **passed** |
| G4 | the guard look-ahead window shrinks to nothing | `EveryVersionRead_FailsItsStep…` | RED (4 failed) |
| G5 | the bash arm drops its `exit` requirement | the guard theory | RED (1 failed) |

**I3 and G3 are the rows that matter, and they are the same shape.** With the matcher accepting every path, the repository scan passes; with the guard detector crediting every line, the guard scan passes. Both scans are vacuous on their own. Only the theories' known-bad rows discriminate, which is why they exist.

**G5 is why the bash guard arm is narrow.** `[ -n "` on its own is not a guard — the workflows already spell that test for unrelated conditions, so one sitting near a read would be credited as protecting it. Dropping the `exit` requirement reds the theory.

**I5b is the honest bound.** Crippled glob plus all three floors deleted is 17/17 green having examined nothing. The floors are the only thing between a crippled glob and a green suite; I predicted I5 would show that with one test's floors and was wrong, because `EveryShippedTextReadPattern…` independently floors the same glob.

**I6 changes no answer today, and that is stated rather than dressed up.** Over the six declarations this replaced, a grep for the value found 4 (the issue's original count), a substring count of the angle-bracket spelling found 12, and the element walk found 6 — three methods, three answers. After the change the substring count and the walk both return 1. The walk is kept for the disagreement it had, not one it has.

## An existing guard caught two things this broke

`CrossAppGuardCiGateTests` (in `Lite.Tests`) failed the first full CI run, and both failures were real.

**The `darling` filter did not reach Lite's project file.** `ProductVersionDeclarationTests` enumerates every csproj in the tree, so adding a version property back to `Lite/PerformanceMonitorLite.csproj` has to run the Darling guard that would catch it — and the filter reached `Lite/**/*.cs` and `Lite/**/*.xaml` but no project file. Fixed with `Lite/**/*.csproj`, derived rather than the single path, matching the two entries above it. The guard has a sanctioned exemption list for this situation, and it is deliberately *not* used here: its own stated rationale is "nothing narrower than `Lite.Tests/**` would reach the sweep anyway", which does not apply to a path a filter can name precisely.

**The absence of `Directory.Build.props` was itself a pin.** `ThePopulationReachesTheProjectItems_AndNotOnlyTheCSharp` asserted MSBuild's `DirectoryBuildPropsPath` / `DirectoryBuildTargetsPath` / `DirectoryPackagesPropsPath` as `["", "", <packages>]`, with a comment explaining that asserting a `Directory.Build.props` existed "would be asserting a fiction". It is not a fiction any more. All three are now asserted by value, and the empty middle entry is the assertion that there is still no `Directory.Build.targets`. Verified against the instrument the test itself uses — `dotnet msbuild -getProperty:DirectoryBuildPropsPath` on both test projects returns exactly the three values now asserted.

Both fixes were proven red-first individually, by compiling the **real** `CrossAppGuardCiGateTests.cs` into a `net10.0` harness (13 tests): reverting the filter entry alone reds `EveryCrossAppSourceRead_IsReachableByTheFilterThatGatesItsSuite`; reverting the assertion alone reds `ThePopulationReachesTheProjectItems_AndNotOnlyTheCSharp`; with both in place, 13/13.

This is the same failure class as the issue, one level out: a pin whose population grew past the filter that runs it. Worth noting the first CI run also showed the ordering hazard — `Run Lite tests` failing aborted the job before `Run Darling tests`, so the new pins did not execute at all on that run.

## Run against the base tree

The strongest check available, and stronger than any single mutation: dropping the pin file alone onto an otherwise unmodified `91c0a919a` fails **4 of 17**, and the failure text is the issue's own finding, independently derived —

- `ExactlyOneMsBuildProperty…` lists all **six** declarations with values, 4 at 3.6.0 and 2 at 3.3.0. It reproduces the corrected count, not the issue body's four.
- `NoProjectHandSetsADerivedVersionProperty` lists all nine hand-set derived properties.
- `EveryVersionRead_NamesTheDeclarationFile` lists all **nine** reads across five files, resolving the gate's `$path` indirection to the literal it falls back to.
- `EveryShippedTextReadPattern…` fails on the absent declaration file.

## Open questions from #3222, and the arm taken

**Do the deprecated projects join? Yes.** `<Version>` describes the binary being built, not a historical release. A binary built from today's tree stamped 3.3.0 is the lie, not the fix — its shared dependencies moved across three releases. CI builds these on release events, and the release gate read one of them, so their version was never inert.

**Does `check-version-bump.yml` read the central file? Yes, and it is forced.** After centralising, no project's XML contains a `<Version>` element, so an XML read of one returns nothing. The main-side read keeps a fallback to `deprecated/Dashboard/Dashboard.csproj`, because `main` does not carry `Directory.Build.props` until the release that introduces it lands there — verified against `main`, which also **does not** have the `Dashboard/Dashboard.csproj` the old chain's second fallback named, so that dead arm is removed as its own comment said to. The docs-only skip logic is untouched and still mirrors `build.yml`'s filter.

**Does `nightly.yml` change? Yes, both of its reads** — `:114` (pwsh, Windows artifacts) and `:473` (bash, the linux job the issue does not mention). Both resolve; the bash one is covered by the pattern-agreement pin above.

## Not verifiable on this PR

`check-version-bump.yml` is `on: pull_request: branches: [main]`, so it does not execute on a PR to `dev`. Its change is validated by the pins and by reading, not by a live run. It fires on the next release PR.

There is no PowerShell on the machine this was written on, so the four pwsh reads are verified by the equivalent .NET XML select inside the pin, not by executing PowerShell. CI executes them; `build.yml`'s now prints its value, so the resolved version is visible in the log from this PR onward.

The `.cmd` scripts are not exercised by CI at all. `build-lite.cmd`'s change is a path swap plus a guard, read but not run.

## CHANGELOG entry text

Not applied here, per lane convention — for the coordinator to append to `[Unreleased]`:

```
- **The product version is declared in exactly one place** ([#3222]) - six csproj files each carried their own `<Version>`, two of them three releases behind, and the release gate read a different file from the one the nightly named artifacts with, so bumping the gate's file shipped artifacts carrying the old version while bumping the nightly's file failed the gate on a correct build. Nothing compared them, and no test had ever read a workflow file. One `<Version>` now lives in `Directory.Build.props` and all 22 projects inherit it. The three deprecated projects were also still hand-setting `AssemblyVersion` / `FileVersion` / `InformationalVersion`, which #2113 deleted from the shipped three because a hand-set copy wins the stamp - that is why `deprecated/Dashboard/Dashboard.csproj` declared 3.6.0 while its binary stamped 3.3.0.0, in the very file the release gate took its judgement from. All nine version reads across the workflows and the root build scripts now name the one declaring file, two of which had been pointing at a path that stopped existing in #1612. Each of those reads now fails its own step when it reads nothing, which a pull request could not previously reveal because every consumer of the value was gated on the release event. Two existing guards needed following: the `darling` path filter now reaches Lite's project file, since the new pin enumerates it, and `CrossAppGuardCiGateTests` had pinned the ABSENCE of `Directory.Build.props` and now asserts all three imported build files by value. A release bump is now one line, and `ProductVersionDeclarationTests` fails the build on a second declaration, a re-added derived property, a reader naming another path, an unguarded reader, or a comment that pushes the text-scraping readers off the element.
```

### CODE REVIEW

File: build-dashboard.cmd [hunk 1]
```diff
@@ -8,8 +8,12 @@ echo  Building Full Edition (Dashboard + Installers)
...
 echo.
 
-:: Get version from csproj
-for /f %%a in ('powershell -Command "([xml](Get-Content Dashboard\Dashboard.csproj)).Project.PropertyGroup.Version | Where-Object { $_ }"') do set VERSION=%%a
+:: Get version from the one file that declares it (#3222)
+for /f %%a in ('powershell -Command "([xml](Get-Content Directory.Build.props)).Project.PropertyGroup.Version | Where-Object { $_ }"') do set VERSION=%%a
+if "%VERSION%"=="" (
+    echo ERROR: Could not determine version from Directory.Build.props.
+    exit /b 1
+)
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply build-dashboard.cmd:16]: Claude posting for Erik Darling

Correct, and you found a third one — the `copy "Installer\bin\Release\..."` at line 65, which was not in the report that reached me. So `build-dashboard.cmd` carries **three** pre-#1612 references, not two.

Deliberately out of scope here, and being resolved by deletion rather than repair in a separate PR: `build-dashboard.cmd`, `build-all.cmd` and `package-release.cmd` are all being removed. The evidence for deleting rather than fixing, verified on `dev`:

- **Nothing references them but each other.** A scan of every `.md`, `.yml` and `.cmd` in the repository finds only `build-all.cmd:14` and `package-release.cmd:23` naming them. No documentation describes any of the three.
- **CI does not invoke them** — `build.yml` publishes directly with `dotnet publish`.
- **None of the four root scripts builds Darling at all.** `build-all.cmd` announces "Building All Performance Monitor Projects" and runs Dashboard then Lite. Repairing the paths would leave a script whose name is wrong about the flagship product, and `package-release.cmd` would keep producing a "release" containing no Darling artifacts.
- Dashboard and the CLI Installer have been deprecated since v3.3.0 and ship no release artifacts.

`build-lite.cmd` survives — it works and it builds something that ships.

**Your observation about the interaction is the sharpest part and worth recording**, because it is a real cost of this PR taken alone: the version-read guard now resolves correctly against `Directory.Build.props`, so the script gets *past* version detection and fails later at `dotnet publish` with a generic project-file-not-found error. This change makes the failure **later and less informative** than it was. That is acceptable only because the file is being deleted; if it were staying, the right order would have been paths first, then the version read.

Also fixed in that same PR, since it is the same defect class one document over: `CONTRIBUTING.md:74` and `:80` document `Dashboard/Dashboard.csproj` and `Installer/PerformanceMonitorInstaller.csproj` without the `deprecated/` prefix, while `README.md:653-654` spells both correctly. A new contributor following CONTRIBUTING gets a failing command on their first build.
